plugins/BEdita/Core/src/Model/Action/DeleteEntityAction.php

```diff
@@ -43,6 +44,13 @@ protected function initialize(array $data)
      */
     public function execute(array $data = [])
     {
+        // If object is not deletable returns 403 FORBIDDEN
+        if ((\BEdita\Core\Model\Table\RolesTable::ADMIN_ROLE === $data['entity']['id'] && $data['entity'] instanceof \BEdita\Core\Model\Entity\Role)
+                ||
+            (\BEdita\Core\Model\Table\UsersTable::ADMIN_USER === $data['entity']['id'] && $data['entity'] instanceof \BEdita\Core\Model\Entity\User)) {
+            throw new ForbiddenException(__d('bedita', 'Could not delete "{0}" 1', $data['entity']));
```

Comment:
Actions are supposed to be a bit more agnostic, IMHO.

I'd prefer to see this model-specific logic in `UsersTable::beforeDelete()` and `RolesTable::beforeDelete()`, and raise a generic exception from there. Then, in `UsersController` and `RolesController` catch that exception and turn it into a specific HTTP error by raising the ForbiddenException.

Reply:
Fine for me, but: `beforeDelete` is not called when using `deleteAll`. Is it ok for us or do we use `deleteAll` sometimes?

Reply:
we could probably disable `deleteAll` or override it with a condition like `id != 1`?

Reply:
`deleteAll` and `updateAll` must be used with extreme care anyway. No lifecycle event is triggered with `deleteAll` or `updateAll` — not only `Model.beforeDelete` — and therefore system corruption may happen in many other ways: cache not invalidated, counters not updated, permissions not checked, …

So far, APIs do not expose ways to delete multiple objects at once — except for relationships with junction tables, whose rows are actually entities, but those are already treated with proper care — and I'd prefer not to do such in any near future. Then, our concerns should be more focused on shell scripts or application-specific logic.

That having been said, `deleteAll` may come in handy at times. We already use it sometimes in tests, where we want a table to be basically `TRUNCATE`d to test empty responses or ensure that the behavior is correct even when data is missing. So I'd prefer to keep it enabled, and give the developers the responsibility of handling their data with proper care. The system is not going to be completely bulletproof anyway, so if somebody _really_ wants to screw up their own database, they will eventually succeed. 🙃